docs/devsite-help/grpc-streaming.md

```diff
@@ -0,0 +1,130 @@
+
+- `WriteAsync` and `TryWriteAsync` send a request
+- `CompleteAsync` and `TryCompleteAsync` are called by the client to indicate that it has
+  finished sending requests
+
+The requests are stored in a buffer until they can be sent; the buffer capacity can
+be specified in the `BidirectionalStreamingSettings` passed into the initial call.
+
+The difference between the methods with a `Try` prefix and those without is about error
+handling:
+
+- A call to `WriteAsync` will throw an exception if the client stream has already been
+  completed, or if the buffer is full. A call to `TryWriteAsync` in the same failure
+  scenarios will return a null `Task` to indicate failure instead.
+- A call to `CompleteAsync` will throw an exception if the client stream has already been
+  completed. A call to `TryCompleteAsync` will return a null `Task` to indicate failure
+  instead.
+
+In most cases, it's more appropriate to use `WriteAsync` and `CompleteAsync`; the `Try` variants
+are provided for situations where you may have multiple threads sending requests and potentially
+completing the stream.
+
+Unlike the "raw" gRPC stub, users do not need to ensure that only a single request is in-flight
+at a time: the buffer takes care of handling multiple requests, so long as the buffer capacity
+is not exceeded. The tasks returned by the above methods complete when the request has been
+written to the underlying stream (or when the stream has actually been marked as completed,
+for `TryCompleteAsync` and `CompleteAsync`).
+
+It's important to complete the request stream, in order to complete the call cleanly.
+
+The sample below demonstrates the use of bidirectional streaming for
+`BigQueryWriteClient.AppendRows`; it deliberately doesn't go into the detail of how the requests are
+created or how the responses are handled, as those are API-specific. Even the aspect of
+"the requests can all be created independently from the responses" is API-specific, but the sample
+provides a reasonably common pattern.
+
+[!code-cs[](../examples/help.Streaming.txt#BidirectionalStreaming)]
+
+## Disposal
+
+While unary RPCs require no special handling, streaming RPC calls *must* be disposed to
+avoid resource leaks. The types representing the calls returned by the initial method call
+(e.g. `ReadRowsStream` and `AppendRowsStream`) implement `IDisposable`, and should be handled
+with a `using` statement where possible. (In more complex scenarios where the call needs to be
+preserved across multiple methods, it should be disposed directly at the end.)
+
+Additionally, users should *attempt* to read all the responses from both server-streaming
+and bidirectional-streaming calls, unless they have particular reasons not to do so. Disposing
+of the call before all the responses have been read effectively aborts the call, which may have
+performance implications depending on the exact timing involved. On the other hand, there
```

Comment:
I would briefly explain these performance implications.

The one for consuming responses you are not interested in are clearer.

Reply:
It's deliberately a bit vague. This is what James wrote in https://github.com/grpc/grpc-dotnet/issues/2086:

> The context behind encouraging people to gracefully complete a request is compared to the alternative of the client or server aborting one side. For example, the client disposing the gRPC call. Aborting can cause performance issues, and even cause the HTTP/2 connection to go into a bad state and be aborted when combined with heavy load.

I didn't want to go into that much detail. But yes, the vagueness is a bit annoying. Thoughts welcome.

Reply:
I think a link to this MS doc is enough ("see here for more details") or something like that.

Reply:
Yup, a link works for me, thanks.